### ISSUE TITLE
Enhancement: Add grid search functionality

### ISSUE THREAD

**Issue by mgiannini** [haifengl/smile#219]
This is an enhancement request to add built-in support for grid search to find optimal parameters for regression and classification models. The python library scikit-learn has a function called [GridSearchCV](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html#sklearn.model_selection.GridSearchCV) which could be used as a reference when considering features features.

**Comment by haifengl** [comment]
This is a little bit harder than it looks like. For a dynamic language Python, it is straightforward to pass the grid of parameters (as a dict) and the trainer will pick the hyperparameters by name. But it is impossible for a static language such as Java and Scala. Even with reflection, it is pretty much impossible as the reflection cares about the data types of parameters, not the parameter name.

A word around is that the caller pass a grid of Trainer objects. However, if the caller already generates this grid, the GridSearchCV is just a loop on the grid. I don't see much values.

**Comment by JohnKoumarelas** [comment]
On a similar matter, is it safe to do this grid search (specifically I want to do it for SVMs) in parallel? Can we assume this for all/some classifiers? I am trying to use parallelStream (Java 8+) for this reason.

**Comment by haifengl** [comment]
Yes.

**Comment by alanmarazzi** [comment]
I am working to Clojure wrappers for SMILE (I just started) and this functionality might be somewhat easier to implement with Clojure. Let me know if this is interesting for you

**Comment by haifengl** [comment]
@alanmarazzi I am not familiar with Clojure. It is easier indeed to implement this in a dynamic language. My worries is how easy Java/Scala programs call it. Besides, I am working on a new design of Classifier interface, which will make grid search possible in plain Java.

**Comment by haifengl** [comment]
@alanmarazzi What's the link to your Clojure project? Would you like to contribute it to the Smile root repo? Thanks!

**Comment by alanmarazzi** [comment]
I have yet to push it, I'm in a very early stage, I wanted to get a better feeling of smile before getting seriously to it.

Anyway, it would be very cool to have Clojure wrappers here as well! If you want to have a better feeling how these things look like you can take a look at the wrappers I released for XGBoost https://gitlab.com/alanmarazzi/clj-boost

I might be able to push something in a couple of weeks (a very partial coverage considering the size of smile), I'll let you know something here as soon as possible.

**Comment by fanweihua** [comment]
Any update on this?

**Comment by haifengl** [comment]
Random search and grid search are supported by `smile.validation.Hyperparameters`. An example is 
```
import smile.io.*;
import smile.data.formula.Formula;
import smile.validation.*;
import smile.classification.RandomForest;

var hp = new Hyperparameters()
    .add("smile.random.forest.trees", 100)
    .add("smile.random.forest.mtry", new int[] {2, 3, 4})
    .add("smile.random.forest.max.nodes", 100, 500, 50);


var train = Read.arff("data/weka/segment-challenge.arff");
var test = Read.arff("data/weka/segment-test.arff");
var formula = Formula.lhs("class");
var testy = formula.y(test).toIntArray();

hp.grid().forEach(prop -> {
    var model = RandomForest.fit(formula, train, prop);
    var pred = model.predict(test);
    System.out.println(prop);
    System.out.format("Accuracy = %.2f%%%n", (100.0 * Accuracy.of(testy, pred)));
    System.out.println(ConfusionMatrix.of(testy, pred));
});
```

**Comment by rayeaster** [comment]
> Random search and grid search are supported by `smile.validation.Hyperparameters`. An example is
> 
> ```
> import smile.io.*;
> import smile.data.formula.Formula;
> import smile.validation.*;
> import smile.classification.RandomForest;
> 
> var hp = new Hyperparameters()
>     .add("smile.random.forest.trees", 100) // a fixed value
>     .add("smile.random.forest.mtry", new int[] {2, 3, 4}) // an array of values to choose
>     .add("smile.random.forest.max.nodes", 100, 500, 50); // range [100, 500] with step 50
> 
> 
> var train = Read.arff("data/weka/segment-challenge.arff");
> var test = Read.arff("data/weka/segment-test.arff");
> var formula = Formula.lhs("class");
> var testy = formula.y(test).toIntArray();
> 
> hp.grid().forEach(prop -> {
>     var model = RandomForest.fit(formula, train, prop);
>     var pred = model.predict(test);
>     System.out.println(prop);
>     System.out.format("Accuracy = %.2f%%%n", (100.0 * Accuracy.of(testy, pred)));
>     System.out.println(ConfusionMatrix.of(testy, pred));
> });
> ```

is this already on master branch or not?

**Comment by haifengl** [comment]
It is released with 2.4.0.
